### PR TITLE
IRMA manual

### PR DESCRIPTION
**Pull request by woutslakhorst**
Some details might be missing, but we'll know when it's used.

### CODE REVIEW

File: docs/pages/getting-started/5-irma-contract.rst [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+It also connects a user to the care organization that user is currently working for.
+The signed contract can be seen as a network enabled session token and should therefore be stored alongside the user session.
+The contract is required for every request that results in personal and/or medical data being retrieved.
+
+Basic requirements
+******************
+
+To use IRMA as a means for signing a contract, the following is required:
+
+- the user has the IRMA app installed on an android or iOS device with camera and an internet connection.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:17]: Android

Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:17]: user also must have added the required attributes to the app

File: docs/pages/getting-started/5-irma-contract.rst [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+The XIS/ECD will have to provide two endpoints for the frontend. One endpoint to start a session and one to get the session result.
+More info on these endpoints will be provided further down.
+
+
+Configuring the Nuts node
+*************************
+
+In the contract signing flow, the device running the IRMA app communicates with the Nuts node directly.
+Therefore the Nuts node needs to be accessible to the public internet.
+All APIs on the Nuts node starting with ``/public`` must be accessible over https without any security measures in place.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:34]: "over HTTPS without any additional security measures that could prevent access by mobile devices."

(HTTPS is a security measure, and a vendor could add additional security measures as long as they don't hamper the functionality)

File: docs/pages/getting-started/5-irma-contract.rst [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+A domain must also be available which resolves to those APIs.
+The domain must be configured on the Nuts node:
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:36]: I think the postfix slash MUST be omitted, it didn't work at first when my public url ended with `/`. Maybe mention this (or fix it in the nuts node).

File: docs/pages/getting-started/5-irma-contract.rst [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+.. code-block:: yaml
+
+  auth:
+    publicurl: https://example.com
+
+The Nuts APIs used for signing will embed this URL in the QR code shown to the user.
+The javascript in the frontend will also use this domain (exposed via the QR code) to check the status of the signing session.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:44]: this domain -> this URL

File: docs/pages/getting-started/5-irma-contract.rst [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+Therefore the domain which serves the frontend must be able to do requests to that domain.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:45]: This depends on the setup. If a proxy serves it all from the same domain this isn't the case.

File: docs/pages/getting-started/5-irma-contract.rst [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+    http:
+        default:
+            cors:
+                origin: "other.com"
+
+Where *other.com* is the domain serving the frontend. For development purposes ``*`` is also allowed.
+If the public APIs are mounted on a different port/interface in the nuts config then the ``default`` key should be changed to ``public`` in the example above.
+
+The XIS/ECD communicates with the Nuts node via the ``/internal`` APIs.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:59]: why is this relevant and what do you expect reader to do with this information?

File: docs/pages/getting-started/5-irma-contract.rst [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+
+Setting up the frontend
+***********************
+
+For the frontend we'll be using the `irma-frontend-packages <https://github.com/privacybydesign/irma-frontend-packages>`_ JS library.
+More info on how to use this library can be found on `https://irma.app/docs/irma-frontend/`_.
+You can choose to load the IRMA frontend packages javascript via a html tag, in which case you'll need to build the JS file yourself given the instructions on `https://github.com/privacybydesign/irma-frontend-packages`_ or you can choose to use ``npm``:
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:66]: > a html tag

"an HTML tag"

Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:66]: "JS" or "javascript"?

File: docs/pages/getting-started/5-irma-contract.rst [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+        default:
+            cors:
+                origin: "other.com"
+
+Where *other.com* is the domain serving the frontend. For development purposes ``*`` is also allowed.
+If the public APIs are mounted on a different port/interface in the nuts config then the ``default`` key should be changed to ``public`` in the example above.
+
+The XIS/ECD communicates with the Nuts node via the ``/internal`` APIs.
+
+Setting up the frontend
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:61]: Should we add a link to Demo EHR repo for convenience?

File: docs/pages/getting-started/5-irma-contract.rst [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+The contents of the ``start`` object configures the initial request to start a signing session. You can control the type of request and the contents.
+In this case, some data from the frontend is sent as JSON. This is optional and no particular data is required.
+The ``url``, in this case ``/web/auth``, must be set in such a way that two backend services are available to the frontend.
+
+.. code-block::
+
+    <url>/session
+    <url>/session/<sessionToken>/result
+
+must both be available on the backend. For the example above this means that both ``/web/auth/session/`` and ``/web/auth/session/<sessionToken>/result`` are available. The ``<sessionToken>`` is the token that will be returned by the call to ``<url>/session/``.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:194]: "must both" is missing something in front of it

File: docs/pages/getting-started/5-irma-contract.rst [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+      mapping: {
+        sessionPtr:      r => r.sessionPtr.clientPtr,
+        sessionToken:    r => r.sessionID
+      }
+    }
+
+The ``session`` object contains all the technical parts to connect the IRMA JS library to your backend.
+The contents of the ``start`` object configures the initial request to start a signing session. You can control the type of request and the contents.
+In this case, some data from the frontend is sent as JSON. This is optional and no particular data is required.
+The ``url``, in this case ``/web/auth``, must be set in such a way that two backend services are available to the frontend.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:187]: this is a complex way to say that everything under `<url>/session` must be accessible by the frontend?

File: docs/pages/getting-started/5-irma-contract.rst [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+      "language": "NL",
+      "version": "v3",
+      "legalEntity": "did:nuts:90348275fjasihnva4857qp39hn",
+      "validFrom": "2006-01-02T15:04:05+02:00",
+      "validDuration": "2h"
+    }
+
+The ``type`` must be one of the valid Nuts contract types, currently only ``BehandelaarLogin`` for Dutch and ``PractitionerLogin`` for English are supported.
+The `language`` selects the correct language, ``NL`` for Dutch and ``EN`` for english. The ``version`` must be v3.
+The legalEntity must refer to the DID of the current organization. The user either selects an organization to login for or is already logged in.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:239]: ``legalEntity``

Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:239]: "to login for, or" (add comma)

for or reads comical

File: docs/pages/getting-started/5-irma-contract.rst [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+From that context the organization must have a DID as described in :ref:`Getting Started on customer integration <connecting-crm>`.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:240]: "From that context"?

File: docs/pages/getting-started/5-irma-contract.rst [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+``validFrom`` is a RFC3339 compliant time string. ``validDuration`` describes how long the contract is valid for. Unix compatible duration strings can be used like ``1h`` or ``60m``. The local system timezone is used to format the date and time string.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:241]: I don't think "Unix duration strings" are a thing? In any case add a link.

I'd guess we're using Golang's `Duration` format; https://golang.org/pkg/time/#ParseDuration

File: docs/pages/getting-started/5-irma-contract.rst [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    GET <url>/session/<sessionToken>/result
+
+where ``<url>`` is the base url configured under ``session.url`` in the javascript options and ``<sessionToken>`` is the token returned by the previous call.
+The backend must implement this API, the implementation must call the following API on the Nuts node:
+
+.. code-block::
+
+    GET /internal/auth/v1/signature/session/<sessionToken>
+
+Any errors in calling this service need to be relayed to the frontend. This will instruct the user on why things went wrong and what to do next.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:290]: "Any error" or "Errors"

File: docs/pages/getting-started/5-irma-contract.rst [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+        "status": "completed",
+        "verifiablePresentation": {
+            ...
+        }
+    }
+
+The ``status`` field has a different content when a different signing means is used.
+The presence of the ``verifiablePresentation`` in the result is the main method of checking if the signing session succeeded.
+``verifiablePresentation`` is the cryptographic proof that needs to be stored in the user session.
+It's required in the oauth flow for obtaining an access token.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:305]: OAuth

File: docs/pages/getting-started/5-irma-contract.rst [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+	    "pbdf.sidn-pbdf.email.email": "tester@example.com"
+      },
+      "credentials": {
+        "organization": "CareBears",
+        "validFrom": "2006-01-02T15:04:05+02:00",
+        "validTo": "2006-01-02T17:04:05+02:00"
+      }
+    }
+
+The ``validity`` will indicate its validity.
```
Comment (reinkrul) [inline docs/pages/getting-started/5-irma-contract.rst:343]: just to be sure: if expired, validity = false?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: docs/pages/getting-started/5-irma-contract.rst (hunk 2 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:34]: the footnote might be a bit too difficult for a getting started. I think it's best to say that it's plain https

Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:34]: Replaced sentence with your suggestion

File: docs/pages/getting-started/5-irma-contract.rst (hunk 3 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:36]: I'll add the mention just to make it clear. Common practice is to not use a trailing `/` in nginx/apache/etc but you never know....

File: docs/pages/getting-started/5-irma-contract.rst (hunk 5 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:45]: Then the condition still holds ;)

File: docs/pages/getting-started/5-irma-contract.rst (hunk 6 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:59]: is out of place, removed it.

File: docs/pages/getting-started/5-irma-contract.rst (hunk 7 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:66]: chose javascript

File: docs/pages/getting-started/5-irma-contract.rst (hunk 8 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:61]: maybe later, not when it's incomplete.

File: docs/pages/getting-started/5-irma-contract.rst (hunk 9 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:194]: it referred to the code block, but I made a sentence just for you ;)

File: docs/pages/getting-started/5-irma-contract.rst (hunk 11 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:239]: arrrrr

File: docs/pages/getting-started/5-irma-contract.rst (hunk 12 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:240]: removed those words.

File: docs/pages/getting-started/5-irma-contract.rst (hunk 13 above)
Comment (woutslakhorst) [reply docs/pages/getting-started/5-irma-contract.rst:241]: I added all valid time units.

**Review by reinkrul — APPROVED** [review]

**Pull request merged**